c7n/deprecated.py

```diff
@@ -0,0 +1,246 @@
+# Copyright 2020 Cloud Custodian Authors.
+# Copyright The Cloud Custodian Authors.
+# SPDX-License-Identifier: Apache-2.0
+
+# Initial thinking around the deprecation is identifying changes in filters and
+# actions. These are likely to be the most common aspects:
+#  * renaming a field
+#  * making an optional field required
+#  * removing a field
+#
+# Examples:
+#  - renaming a filter itself
+# c7n_azure/resources/key_vault
+# @KeyVault.filter_registry.register('whitelist')
+# - filter 'whitelist' has been deprecated (replaced by 'allow') and will be removed after 2021-03-21
+#
+# - mark -> tag
+# - unmark, untag -> remove-tag
+#
+#
+#  - renaming filter attributes
+# c7n/filters/iamaccess
+#    schema = type_schema(
+#        'cross-account',
+#        ...
+#        whitelist_from={'$ref': '#/definitions/filters_common/value_from'},
+#        whitelist={'type': 'array', 'items': {'type': 'string'}},
+#        ...)
+# - filter field 'whitelist' has been deprecated (replaced by 'allow') and will be removed after 2021-03-21
+#
+#
+# c7n/tags.py
+#  - optional attributes becoming required, in this case one of 'days' or 'hours'
+#  - optional action fields deprecated (one of 'days' or 'hours' must be specified) and will become an error after 2021-03-21
+#  - optional action field 'tag' deprecated (must be specified) and will become an error after 2021-03-21
+#
+
+from datetime import datetime
+
+
+def action(name, replacement, removed_after):
```

Comment:
part of me is thinking we just return back a data struct here, and actually instantiate in get_deprecations, to defer instance construction, and also give opportunity to context bind.

Reply:
Interesting... I'm mulling this one over.